### ISSUE TITLE
Black spirit quest

### ISSUE THREAD

**Issue by just4omp** [kajdaq/stunning-chainsaw#12]
Sometimes script dont accept new bs quest.
Script open black spirit, open bs quest, then just close this menu.
Here is log. Last rows, 07:20:47

![Screenshot_20200410-073343](https://user-images.githubusercontent.com/58849548/78962755-c9d95700-7afd-11ea-9606-8201dd1da26d.png)
[log.txt](https://github.com/kajdaq/stunning-chainsaw/files/4459880/log.txt)

**Comment by wrathhunter** [comment]
i know...it's non repetitive..you need to run the script everytime and then close it again and then again start after atleast 10 minutes..

**Comment by wrathhunter** [comment]
https://github.com/kajdaq/stunning-chainsaw/issues/6#issuecomment-611966340

**Comment by just4omp** [comment]
> [#6 (comment)](https://github.com/kajdaq/stunning-chainsaw/issues/6#issuecomment-611966340)

for me its ok, i can leave script for 12+ hours without any serious  loops and bugs (if only ankulua or emulator does not crash). just mentioned today, that script did not accept new quest, btw previous two quests script accepted b itself.

**Comment by wrathhunter** [comment]
how?can you please share the lua folder in zip with the config screenshots please?

**Comment by wrathhunter** [comment]
and also ld player settings and ingame settings

**Comment by just4omp** [comment]
> and also ld player settings and ingame settings

lua standard from ankulua forum, ingame all max except fps, ld default 1080p, root disabled.
zip below
[script.zip](https://github.com/kajdaq/stunning-chainsaw/files/4461709/script.zip)

**Comment by wrathhunter** [comment]
thanks let me test

**Comment by wrathhunter** [comment]
does your ultimate skill fire up every 300 seconds?
because for me it's not working..only worked for once


i have another question..how could you run your script in ankulua with root disabled?

**Comment by just4omp** [comment]
yes, it use it 300s +- (10 - 600s, depends on how much seconds you set in script for random) and if i have enough mana when script try to use)) but sometimes i see how it use ult

I use media projection in ankulua. Its faster and looks like more stable.
![Screenshot_20200410-173404](https://user-images.githubusercontent.com/58849548/78998552-8c061e00-7b51-11ea-97af-eb2d50aee05f.png)

**Comment by wrathhunter** [comment]
alright

**Comment by wrathhunter** [comment]
let me try through media projection..because through root it does not do anything for the second time..only for the first time

**Comment by wrathhunter** [comment]
if we change the random value to 0 and 0...means no random..it will always search..then it should fire up the ulti in every 300 seconds ..right

**Comment by just4omp** [comment]
> if we change the random value to 0 and 0...means no random..it will always search..then it should fire up the ulti in every 300 seconds ..right

i think yes, if 0-0 then without random. its easy to track through log file how often script try to use ult. but i dont know what happen if in same time will trigger ult and for example absorb black energy.

**Comment by wrathhunter** [comment]
ok let me try

**Comment by wrathhunter** [comment]
it only searches for them but don't trigger them..
it's really embarassing..
for the first time it searches and triggers..
next it only searches them but don't trigger them

**Comment by wrathhunter** [comment]
![Screenshot (66)](https://user-images.githubusercontent.com/48764052/79001674-8964f200-7b6c-11ea-824b-29b48bab5d96.png)
see only finding option..no trigger

**Comment by just4omp** [comment]
> see only finding option..no trigger

if i understand right it find and trigger (try to use) it, i have same lines in my log and its ok (used when ult menu was open, will try now with closed ult menu, but s i remember its ok too). may be some ingame conditions interrupt ult?

**Comment by wrathhunter** [comment]
why will any ingame condition interrupt it?
how it ran for the first time?
if the ult is used once then it will stay open and next time when it comes it finds the menu is on and just leave as it is without triggering..

but if i deselect all  the options in the script except the ulti one..then the ulti works perfectly fine..

that means the script has some error which prevent from triggering the next iteration ..

**Comment by wrathhunter** [comment]
function doUltimateSkills()
    if(not shouldUseUltimateSkills or not enableUltimateSkills) then return end

    writeLog("[Ultimate Skills] Starts")

    local usMenu = {target =  Pattern("btn_ultimate-skills_.png"):similar(0.8), region = Region(1229, 348, 50, 56) }
    local us1 = {target =  Pattern("btn_ultimate-skills-button_.png"):similar(0.7), region = Region(1038, 339, 81, 78) }
    local us2 = {target =  Pattern("btn_ultimate-skills-button__1.png"):similar(0.7), region = Region(1123, 339, 75, 77) }
    local us3 = {target =  Pattern("btn_ultimate-skills-button__2.png"):similar(0.7), region = Region(1204, 339, 74, 77) }
    
    if(existsClick(usMenu)) then
        writeLog("[Ultimate Skills] Open menu")
        wait(1)
    end
    
    if(existsClick(us1)) then
        writeLog("[Ultimate Skills] Use Skill 1")
        existsClick(us1)
        wait(1)
    end

    if(existsClick(us2)) then
        writeLog("[Ultimate Skills] Use Skill 2")
        existsClick(us2)
        wait(1)
    end

    if(existsClick(us3)) then
        writeLog("[Ultimate Skills] Use Skill 3")
        existsClick(us3)
    end

    periodicallyUSeUltimateSkillsLastTime = os.time(os.date("!*t")) + math.random(randomExtraTime[0], randomExtraTime[1])
    shouldUseUltimateSkills = false
    wait(1)

    writeLog("[Ultimate Skills] Done")
    writeLog("----------------------------------------------------------------")
end








now here i found the reason why it does not do ultimate quests

local usMenu = {target =  Pattern("btn_ultimate-skills_.png"):similar(0.8), region = Region(1229, 348, 50, 56) }

this is the ultimate skill opening image..
![btn_ultimate-skills_](https://user-images.githubusercontent.com/48764052/79009426-06986300-7b7d-11ea-99b5-55d1ccc31e93.png)

 if(existsClick(usMenu)) then
        writeLog("[Ultimate Skills] Open menu")
        wait(1)
    end

this code checks for this image..

for the first time it checks and opens the skill bar
now when it comes next it searches for the exact image but till then it is already opened
so it does not find the image and does not do the ultimate skills..

so now if we close the bar each time after it executes all the skills will solve the problem or else we need to expand the search region so that it will search that and execute

**Comment by wrathhunter** [comment]
periodicallyUSeUltimateSkillsLastTime = os.time(os.date("!*t")) + math.random(randomExtraTime[0], randomExtraTime[1])


in this line also we can just remove the math.rand function and give the actual amount for the 2 ultimate skills
first one takes 2 minutes and second one takes 3 minutes..

so we can give os.time(os.date("!*t")) + 120 && os.time(os.date("!*t")) + 180

so that it will trigger them as soon as it is available

**Comment by wrathhunter** [comment]
i can edit the script but i need the latest version..shawn davo gave me a older version which is not so helpfull..
if @kajdaq could share it here then whose ever is good enough to make the script bug free he can try..

actually we all will make this completely bug free and then we will add some more features
like new high hp potion and tower training etc..

**Comment by wrathhunter** [comment]
by the way i made my script with winautomation which is completely hard coded...no image recognition is there..so if anybody need then they can make too..atleast just for the timebeing till our dev solves all these issues
